homeassistant/components/noonlight/__init__.py

```diff
@@ -0,0 +1,195 @@
+
+    @property
+    def access_token_expiry(self):
+        """Return the timestamp when the access token expires."""
+        return self.config \
+            .get(CONF_ACCESS_TOKEN, {}) \
+            .get('expires', dt_util.utc_from_timestamp(0))
+
+    @property
+    def access_token_expires_in(self):
+        """Returns the timedelta when the token expires."""
+        return self.access_token_expiry - dt_util.utcnow()
+
+    @property
+    def should_token_be_renewed(self):
+        """Returns true if the token needs to be renewed."""
+        return self.access_token is None \
+            or self.access_token_expires_in <= self._time_to_renew
+
+    @property
+    def client(self):
+        if self._client is None:
+            import noonlight as nl
+            self._client = nl.NoonlightClient(token=self.access_token)
+            api_url = self.config.get(CONF_API_ENDPOINT)
+            self._client._base_url = api_url
+        return self._client
+
+    async def check_api_token(self, force_renew=False):
+        _LOGGER.debug("checking if token needs renewal, expires: {0:.1f}h"
+                      .format(self.access_token_expires_in
+                              .total_seconds() / 3600.0))
+        if self.should_token_be_renewed or force_renew:
+            try:
+                _LOGGER.debug("Renewing Noonlight access token...")
+                path = self.config.get(CONF_TOKEN_ENDPOINT)
+                data = {
+                    'id': self.config.get(CONF_ID),
+                    'secret': self.config.get(CONF_SECRET)
+                }
+                token_response = await self.client._post(path=path, data=data)
+                if 'token' in token_response and 'expires' in token_response:
+                    self._set_token_response(token_response)
+                    _LOGGER.debug("token renewed, expires at {0} ({1:.1f}h)"
+                                  .format(self.access_token_expiry,
+                                          self.access_token_expires_in
+                                              .total_seconds()/3600.0))
+                    self.hass.bus.fire(EVENT_NOONLIGHT_TOKEN_REFRESHED, {
+                            'token': self.client._token
+                        })
+                    return True
+                raise Exception("unexpected token_response: {}"
+                                .format(token_response))
+            except Exception:
+                _LOGGER.exception("Failed to renew Noonlight token!")
+                return False
+        return True
+
+    def _set_token_response(self, token_response):
+        expires = dt_util.parse_datetime(token_response['expires'])
+        if expires is not None:
+            token_response['expires'] = expires
+        else:
+            token_response['expires'] = dt_util.utc_from_timestamp(0)
+        self.client.set_token(token=token_response.get('token'))
+        self.config[CONF_ACCESS_TOKEN] = token_response
```

Comment:
We probably shouldn't modify the config directly. I'd make a copy first before storing the config as an instance attribute if we need to modify it.